Gradle quarkusDev fails when using Java toolchains

## Describe the bug
When using Gradle Java toolchains quarkus dev is not run with JDK provided in toolchains but it uses current system JDK.
Toolchains reference: https://docs.gradle.org/current/userguide/toolchains.html

For example if my system Java version is 11 and I have defined Java toolchains with version 15, gradle will compile classes with JDK 15, but quarkusDev will start server with JDK 11.

Exception:
```
2021-04-07 19:10:44,011 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:134)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:65)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:42)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:119)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
        ... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:277)
        ... 15 more
Caused by: java.lang.RuntimeException: Failed to initialize Arc
        at io.quarkus.arc.Arc.initialize(Arc.java:26)
        at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:40)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy_0(ArcProcessor$generateResources-1025303321.zig:76)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy(ArcProcessor$generateResources-1025303321.zig:40)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:213)
        ... 15 more
Caused by: java.lang.UnsupportedClassVersionError: org/acme/GreetingResource has been compiled by a more recent version of the Java Runtime (class file version 59.0), this version of the Java Runtime only recognizes class file versions up to 55.0
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:411)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:371)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:398)
        at org.acme.GreetingResource_Bean.<init>(GreetingResource_Bean.zig:47)
        at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Default_ComponentsProvider.zig:209)
        at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Default_ComponentsProvider.zig:38)
        at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:112)
        at io.quarkus.arc.Arc.initialize(Arc.java:20)
        ... 19 more

``` 

### Expected behavior
gradle quarkusDev works without issues.

### Actual behavior
gradle quarkusDev fails with exception.


## To Reproduce
Steps to reproduce the behavior:
1. Run gradle quarkusDev

### Reproducer
[code-with-quarkus-test.zip](https://github.com/quarkusio/quarkus/files/6273386/code-with-quarkus-test.zip)


## Environment (please complete the following information):

### Output of `uname -a` or `ver`
MacOS 10.15.7

### Output of `java -version`
openjdk version "11.0.6" 2020-01-14

### Quarkus version
1.13.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Gradle 6.8.3

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/GradleDevModeLauncher.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/GradleDevModeLauncher.java
@@ -10,16 +10,17 @@ public class GradleDevModeLauncher extends QuarkusDevModeLauncher {
      * Initializes the launcher builder
      *
      * @param logger the logger
+     * @param java java binary to use
      * @return launcher builder
      */
-    public static Builder builder(Logger logger) {
-        return new GradleDevModeLauncher(logger).new Builder();
+    public static Builder builder(Logger logger, String java) {
+        return new GradleDevModeLauncher(logger).new Builder(java);
     }
 
     public class Builder extends QuarkusDevModeLauncher.Builder<GradleDevModeLauncher, Builder> {
 
-        private Builder() {
-            super(null);
+        private Builder(String java) {
+            super(java);
         }
     }
 

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java
@@ -25,14 +25,20 @@ import org.gradle.api.artifacts.ResolvedArtifact;
 import org.gradle.api.artifacts.ResolvedDependency;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.Input;
 import org.gradle.api.tasks.InputDirectory;
+import org.gradle.api.tasks.Nested;
 import org.gradle.api.tasks.Optional;
 import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.SourceSetContainer;
 import org.gradle.api.tasks.TaskAction;
 import org.gradle.api.tasks.compile.JavaCompile;
 import org.gradle.api.tasks.options.Option;
+import org.gradle.jvm.toolchain.JavaLauncher;
+import org.gradle.jvm.toolchain.JavaToolchainService;
+import org.gradle.jvm.toolchain.JavaToolchainSpec;
 
 import io.quarkus.bootstrap.BootstrapConstants;
 import io.quarkus.bootstrap.model.AppArtifact;
@@ -164,6 +170,16 @@ public class QuarkusDev extends QuarkusTask {
         this.compilerArgs = compilerArgs;
     }
 
+    @Nested
+    @Optional
+    public Provider<JavaLauncher> getJavaLauncher() {
+        Project project = getProject();
+        JavaToolchainService toolChainService = project.getExtensions().getByType(JavaToolchainService.class);
+        JavaToolchainSpec toolchainSpec = project.getExtensions().getByType(JavaPluginExtension.class).getToolchain();
+        Provider<JavaLauncher> javaLauncher = toolChainService.launcherFor(toolchainSpec);
+        return javaLauncher;
+    }
+
     @TaskAction
     public void startDev() {
         if (!getSourceDir().isDirectory()) {
@@ -192,7 +208,14 @@ public class QuarkusDev extends QuarkusTask {
 
     private QuarkusDevModeLauncher newLauncher() throws Exception {
         final Project project = getProject();
-        GradleDevModeLauncher.Builder builder = GradleDevModeLauncher.builder(getLogger())
+
+        String java = null;
+        Provider<JavaLauncher> javaLauncher = getJavaLauncher();
+        if (javaLauncher.isPresent()) {
+            java = javaLauncher.get().getExecutablePath().getAsFile().getAbsolutePath();
+        }
+
+        GradleDevModeLauncher.Builder builder = GradleDevModeLauncher.builder(getLogger(), java)
                 .preventnoverify(isPreventnoverify())
                 .projectDir(project.getProjectDir())
                 .buildDir(getBuildDir())

